FC01_Feature:  Audit Logging

yasaminashoori opened the pull request:
# Audit Logging 

**linked issue: https://github.com/ashkanRmk/FastCrud/issues/1**


Hi Mr. Rahmani 🌹  Thank you for the recent updates, it is great and I use FastCrud in my daily works.
 I apologize for the extended development time on this implementation I wanted to ensure it follows the project's architecture patterns correctly and maintains the high quality standards of FastCrud. 

# Features
This PR adds audit logging functionality to FastCrud, enabling automatic tracking of all CRUD operations including Create, Update, Delete on entities. 

###  New Endpoint for Audit

```
app.MapAuditLogs<AuditEntry>(
    "/api/audit-logs",
    tagName: "Audit",
    groupName: "v1"
);
```

### Response Example 
However, the responses need refinement. I want to confirm the implementation on your end, and then, if permissible, I will proceed with some refactoring.

<img width="1733" height="417" alt="image" src="https://github.com/user-attachments/assets/0c31f52a-2e30-4224-ace0-bed7ed763e55" />


# 🔴 Architecture and Patterns Details 

### 1. Capture approach: 
 **EF Core SaveChanges Interceptor  and Manual Service**  (Manual Service for `EfAuditService<TDbContext, TAuditEntry>` )

```
public sealed class EntityAuditingInterceptor<TAuditEntry> : SaveChangesInterceptor
{
    public override InterceptionResult<int> SavingChanges()
    public override ValueTask<InterceptionResult<int>> SavingChangesAsync()
}
```
 
### 2. Persistence: 

 **This code implemented with sync / async approach and audit entries will save immediately and doesn't go to the queue for processing**

```
public override InterceptionResult<int> SavingChanges()

public override ValueTask<InterceptionResult<int>> SavingChangesAsync()

```

# 🔍 Performance and Reliability

### 1. Batching / Deferred
**We used both in the _EntityAuditingInterceptor_ but we don't heve neither in _EfAuditService_**

  ✅EntityAuditingInterceptor 

```
private void AddAuditEntries(DbContext? context)
{
    var auditEntries = new List<TAuditEntry>();
    
    foreach (var entry in context.ChangeTracker.Entries())
    {
        auditEntries.Add(auditEntry);
    }
    
    if (auditEntries.Any())
    {
        context.Set<TAuditEntry>().AddRange(auditEntries); 
    }
}

```

❌ EfAuditService

```
public async Task LogAsync<T>(T entity)
{
    _context.Set<TAuditEntry>().Add(auditEntry); 
    await _context.SaveChangesAsync(cancellationToken); 
}
```

### 2. Transaction coupling

✅EntityAuditingInterceptor  has Same Transaction

```
private void AddAuditEntries(DbContext? context)
{
    context.Set<TAuditEntry>().AddRange(auditEntries);
}

```

 ❌ EfAuditService has Fire-and-forget

```
public async Task LogAsync<T>(T entity)
{
    _context.Set<TAuditEntry>().Add(auditEntry);
    await _context.SaveChangesAsync(cancellationToken);
}

```

❌  CrudServices has Fire-and-forget:

```
async Task<TAgg> CreateAsync(object input, CancellationToken cancellationToken)
{
    await repository.AddAsync(entity, cancellationToken);
    await repository.SaveChangesAsync(cancellationToken);
    
    if (auditService != null)
    {
        await auditService.LogAsync(entity, AuditAction.Create,);
    }
}
```

### 3. Handling large diffs (field-level vs full snapshot)
**We used field-level tracking.**

```
case EntityState.Modified:
    var modifiedProperties = entry.Properties
        .Where(p => p.IsModified && !p.Metadata.IsPrimaryKey()) 
        .ToList();
        
    auditEntry.OldValues = SerializeModifiedValues(entry, entry.OriginalValues, modifiedProperties);
    auditEntry.NewValues = SerializeModifiedValues(entry, entry.CurrentValues, modifiedProperties);
```

```
if (property.Metadata.IsForeignKey() ||
    property.Metadata.IsShadowProperty() ||
    property.Metadata.IsKey()) continue;
```


# 🔍 Configuration

### 1. Enable/disable globally or per module?


✅ **Global Enable exists in _AddEfAuditing registration_**

```
builder.Services.AddEfAuditing<AppDbContext, AuditEntry>();

```

We don't have any disable globally in Configuration Options and we ONLY have per entity in _IAuditEntry_ and _HasId checks_


### 2.  DI Registration Pattern 
 
✅ Yes

Extension method pattern
Generic constraints
coped lifetime 
Factory pattern for IAuditService
Method AddCustomAuditUserProvider 

```
public static IServiceCollection AddEfAuditing<TDbContext, TAuditEntry>(this IServiceCollection services)
{
    services.AddScoped<IAuditUserProvider, DefaultAuditUserProvider>();
    services.AddScoped<EntityAuditingInterceptor<TAuditEntry>>();
    services.AddScoped<IAuditService>(sp => new EfAuditService<x>);
    return services;
}
services.AddCustomAuditUserProvider<MyProvider>();
```

Commit:
Merge branch 'dev'

juniorCoder18 committed:
FC01_Feature_Audit_Logging: add audit

juniorCoder18 committed:
FC01_Feature_Audit_Logging:  improvement and clean code

juniorCoder18 committed:
FC01_Feature_Audit_Logging:  typo mistake

juniorCoder18 committed:
FC01_Feature_Audit_Logging:  typo mistake

yasaminashoori committed:
Merge branch 'ashkanRmk:main' into FC01_Feature_Audit_Logging

ashkanRmk commented:
Hi Yasamin,
Thank you so much for this PR and for using FastCrud in your daily work.
Really appreciate the effort and care you put into aligning with the architecture. 
Audit logging is a great addition.

I’ll review it within the next few days and let you know if anything needs adjustment.

Thanks again for the contribution! 🙌

yasaminashoori commented:
Thanks a lot for the quick feedback🌹 I'm Glad to hear audit logging is a useful addition. Looking forward to your review and will address any adjustments you suggest.

yasaminashoori commented:
Thanks a lot for the quick feedback🌹 I'm Glad to hear audit logging is a useful addition. Looking forward to your review and will address any adjustments you suggest.

ashkanRmk commented:
Heads-up: I just pushed some updates to the dev branch (structural tweaks + minor service changes), so please pull/rebase from dev and resolve any merge conflicts before I start the review. That’ll keep feedback focused on the audit layer itself.

To help speed things up, could you add (or confirm) a concise section in the PR description covering:

Architecture / Pattern

Capture approach: interceptor, decorator, domain events, EF Core SaveChanges hook, or pipeline behavior?
Sync vs async persistence (any queue/outbox pattern planned)?

Performance / Reliability

Any batching or deferred write?
Transaction coupling (in same transaction vs fire-and-forget).
Handling large diffs (field-level vs full snapshot).

Configuration

How to enable/disable globally or per module.
DI registration pattern.

If you drop that summary in, the review will go much faster. Let me know once you’ve rebased and updated.
Great work so far, and thanks again!

yasaminashoori commented:
Thank you for the comprehensive feedback! ❤️ I'll rebase from the dev branch and resolve any conflicts, I updated the descriptions of PR and it includes all of the concerns you raised.

Looking back, I realize I should have initiated a technical discussion and maybe modeling document  before implementing the audit layer to align on the architectural approach and gather your input on design decisions.

I've learned from this experience and will make sure to engage in upfront technical discussions for future significant features. 
Once I've completed the rebase I'd greatly appreciate any additional feedback or suggestions you might have on the implementation approach or areas for improvement.

Thanks again for your patience and thorough guidance 🌹

juniorCoder18 committed:
FC01_Feature_Audit_Logging: resolve conflicts

juniorCoder18 committed:
FC01_Feature_Audit_Logging: resolve conflicts

juniorCoder18 committed:
FC01_Feature_Audit_Logging: resolve conflicts

yasaminashoori commented:
I resolved the conflicts, and I am waiting for your feedback 💯

ashkanRmk commented:
Hi Yasamin ✌🏻,
I’ve reviewed the code and documentation for this pull request, and I wanted to share a few considerations and suggestions that might help further improve the approach:

### Column Type for `OldValues` and `NewValues`:
Using columns like `OldValues` and `NewValues` to store change snapshots is a solid idea. However, since the length of changes can be unpredictable,especially with large entities.
These columns need to be set as `nvarchar(max)`. In practice (particularly in large companies), this can lead to performance issues and substantial memory reservation on disk, which is a concern for scalability and efficiency.

### Entity-Specific Audit Endpoints:
It may be better to provide an audit log endpoint for each entity separately, rather than requiring users to specify the entity name manually. This would make the API more intuitive and easier to consume.

### Configurable CRUD Audit Routes:
Just as other operations are configurable, it would be valuable to offer CRUD audit endpoints only if auditing is enabled for an entity. This way, routes are exposed conditionally, matching user configuration and improving clarity.

---

Overall, this is a great contribution and a strong foundation for auditing in FastCrud. Thank you again for your hard work!
Looking forward to seeing how the feature evolves.

yasaminashoori commented:
Hi Mr. Rahmani,
Thank you so much for taking the time to review and share such detailed feedback  I really appreciate
 it.🌹

I’m really sorry I haven’t had the chance to go through everything yet. It’s been a bit busy these past few days, but I’ll take some time this weekend to review everything carefully and make the updates.

Thanks again for your patience and all the helpful insights!

yasaminashoori committed:
Address maintainer feedback

yasaminashoori committed:
Configure OldValues and NewValues, add endpoints, remove unnecessary dependency

yasaminashoori committed:
resolve merge conflicts

yasaminashoori committed:
formatting

yasaminashoori commented:
Thank you for your great feedback.

**Summary of the last  changes:**

1. Set OldValues and NewValues columns to nvarchar(max) 
2. Added entity specific audit endpoints 
3. Made audit routes configurable with includeEntities and excludeEntities parameters

ashkanRmk commented:
Hi Yasamin

Thank you so much for this contribution, and especially for the detailed PR description and your patience in keeping it updated, the effort really shows, and the interceptor design is genuinely the right approach. I tested the branch locally (built it, ran the sample, and exercised create/update/delete), and before merging I'd like to ask for a few changes. I found some issues that I want to share with you:

1. Every operation is audited twice (main blocker)

Right now the PR has two capture mechanisms running at the same time: the EntityAuditingInterceptor (fires on every SaveChanges) and the manual `IAuditService.LogAsync` calls inside `CrudService`. When I created one customer, I got two audit rows, one from each path, each with a different JSON format (the interceptor writes a camelCase field diff, the manual service writes a full PascalCase entity dump).

👉 Please keep only the interceptor and remove the `IAuditService` calls from `CrudService` (and the double wiring in the sample's `Program.cs`). The interceptor is better in every way: it runs in the same transaction as the operation, it produces clean field-level diffs, and it catches every save — not just the ones going through CrudService.

2. Delete is audited before it actually happens

In `DeleteAsync`, the audit entry is saved before the entity is deleted. If the delete fails afterward, the audit log will claim the entity was deleted when it wasn't — the log would be lying. Removing the manual service (point 1) fixes this automatically, since the interceptor writes the audit row in the same save as the delete.

Smaller things (quick fixes):

The interceptor file name has a trailing space: `EntityAuditingInterceptor .cs` — please rename.
The `AddEfAuditing(Type)` overload doesn't register IAuditQueryService, so MapAuditLogs would fail for anyone using that overload.
The hardcoded count: 100 in the endpoint should be a query parameter.
I'd suggest dropping `IAuditable` (CreatedAt/CreatedBy stamping) from this PR — it's a separate feature, currently unused and untested. It deserves its own PR.

Once the interceptor is the single capture path and the points above are addressed, this will be a really solid feature, the field-level diffing, primary-key handling, and the generic `TAuditEntry` design are all well done. 

Thanks again for the great work and for asking for design feedback before refactoring, that was exactly the right call. Looking forward to the next iteration! 🚀

yasaminashoori commented:
@ashkanRmk 
Hi Mr. Rahmani,  
Thank you for the detailed review. It was really helpful.

I've fixed all the issues you mentioned:
- Removed `IAuditService` completely. Auditing now only happens through `EntityAuditingInterceptor` during `SaveChanges` so there's no duplicate audit anymore.

- This also fixes the delete issue. If a delete fails then no audit log is created
  
- Renamed `EntityAuditingInterceptor .cs` to `EntityAuditingInterceptor.cs`
    
- `AddEfAuditing(Type)` now registers `IAuditQueryService` too
    
- Replaced the hardcoded `100` with `[FromQuery] int count = 100` in the audit endpoints
    
- Removed `IAuditable` and the `CreatedAt`/`CreatedBy` logic from the interceptor since that belongs in a separate PR
    

I also tested everything locally:
- Create generates one audit record per entity for example a`Customer` and a `Tag` which is expected
- Delete is logged correctly in the same transaction and also about uopdate in both order and custoomer
- The `?count=` parameter works on both audit endpoints

Thanks again for the helpful feedback I've learned a lot 🌹
